include/swift/ClangImporter/SwiftAbstractBasicReader.h

```diff
@@ -0,0 +1,86 @@
+//===- SwiftAbstractBasicReader.h - Clang serialization adapter -*- C++ -*-===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2017 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+//
+// This file provides an intermediate CRTP class which implements most of
+// Clang's AbstractBasicReader interface, paralleling the behavior defined
+// in SwiftAbstractBasicWriter.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef SWIFT_CLANGIMPORTER_SWIFTABSTRACTBASICREADER_H
+#define SWIFT_CLANGIMPORTER_SWIFTABSTRACTBASICREADER_H
+
+#include "clang/AST/AbstractTypeReader.h"
+// AbstractBasicReader.h doesn't include what it uses
+#include "clang/AST/DeclObjC.h"
+
+namespace swift {
+
+/// The subclass must implement:
+///   uint64_t readUInt64();
+///   clang::IdentifierInfo *readIdentifier();
+///   clang::Stmt *readStmtRef();
+///   clang::Decl *readDeclRef();
+template <class Impl>
+class DataStreamBasicReader
+       : public clang::serialization::DataStreamBasicReader<Impl> {
+  using super = clang::serialization::DataStreamBasicReader<Impl>;
+public:
+  using super::asImpl;
+  using super::getASTContext;
+
+  DataStreamBasicReader(clang::ASTContext &ctx) : super(ctx) {}
+
+  /// Perform all the calls necessary to write out the given type.
+  clang::QualType readTypeRef() {
+    auto kind = clang::Type::TypeClass(asImpl().readUInt64());
+    return clang::serialization::AbstractTypeReader<Impl>(asImpl()).read(kind);
+  }
+
+  bool readBool() {
+    return asImpl().readUInt64() != 0;
+  }
+
+  uint32_t readUInt32() {
+    return uint32_t(asImpl().readUInt64());
+  }
+
+  clang::Selector readSelector() {
+    uint64_t numArgsPlusOne = asImpl().readUInt64();
+
+    // The null case.
+    if (numArgsPlusOne == 0)
+      return clang::Selector();
+
+    unsigned numArgs = unsigned(numArgsPlusOne - 1);
+    SmallVector<clang::IdentifierInfo *, 4> chunks;
+    for (unsigned i = 0, e = std::max(numArgs, 1U); i != e; ++i)
+      chunks.push_back(asImpl().readIdentifier());
+
+    return getASTContext().Selectors.getSelector(numArgs, chunks.data());
+  }
+
+  clang::SourceLocation readSourceLocation() {
+    // Always read null.
+    return clang::SourceLocation();
```

Comment:
Should this have a `TODO: Also track source locations. That requires Foo Bar.` or are we planning to keep it this way? I remember one of my patches needed a source location to print the Clang diagnostic and we just made up a fake one or something...

Reply:
I don't think there will ever be a meaningful way for us to stably serialize a Clang source location.  Fortunately, they do not generally appear in Clang types, especially in prototypes.

Th serialization/deserialization logic is designed to have a precondition of passing the isSerializable check.  It looks like I'm missing a comment to that effect on the deserialization side.

Reply:
Gotcha'. I think your additional context about 

> I don't think there will ever be a meaningful way for us to stably serialize a Clang source location. Fortunately, they do not generally appear in Clang types, especially in prototypes.

would be useful to have in a comment next to the code, perhaps here itself.